### PR TITLE
[go1.16] Update to go1.16

### PR DESCRIPTION
**Pull request by justaugustus**
**What type of PR is this?**

/kind feature
/area dependency

**What this PR does / why we need it**:

- Update to go1.16

Tracking issue: https://github.com/kubernetes/release/issues/1834
Slack thread: https://kubernetes.slack.com/archives/C2C40FMNF/p1610074100305000
cc: @kubernetes/release-engineering 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
- Kubernetes is now built with Golang 1.16
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by hasheddan — APPROVED** [review]
/lgtm

**Comment by justaugustus** [comment]
/assign @dims 
FYI @nikhita

**Comment by justaugustus** [comment]
(now why did I think this was going to be simple...)
/retest

**Comment by dims** [comment]
@justaugustus is the intention to merge the RC1 or wait for the final release ( which is probably in a few days 🤞🏾 )

**Comment by justaugustus** [comment]
> @justaugustus is the intention to merge the RC1 or wait for the final release ( which is probably in a few days 🤞🏾 )

I was hoping to get the RC1 in, but we'll see what the tests have to say about that.
Depending on how timing lines up, this may just end up being the go1.16.0 PR.

File: staging/publishing/rules.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ recursive-delete-patterns:
 - BUILD.bazel
 - "*/BUILD.bazel"
 - Gopkg.toml
-default-go-version: 1.15.7
+default-go-version: 1.16rc1
```
Comment (justaugustus) [inline staging/publishing/rules.yaml:7]: @dims @nikhita -- Have you seen [this](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-publishing-bot-validate/1355249121295863808) crop up w/ pre-release versions before?

```console
 F0129 20:22:31.932516     163 main.go:35] Invalid rules file "/home/prow/go/src/k8s.io/kubernetes/staging/publishing/rules.yaml": validation errors:
- specified go version 1.16rc1 must be a valid go version: Invalid character(s) found in minor number "16rc1" 
```

Comment (dims) [reply staging/publishing/rules.yaml:7]: @justaugustus we validate using blang/semver, that's what is failing. See snippet https://play.golang.org/p/gNIlfcIoibb

Comment (nikhita) [reply staging/publishing/rules.yaml:7]: Created https://github.com/kubernetes/publishing-bot/pull/245

**Comment by nikhita** [comment]
Hmm, looks like `pull-publishing-bot-validate` is not fetching the new publishing-bot code.

File: staging/publishing/rules.yaml (hunk 1 above)
Comment (nikhita) [reply staging/publishing/rules.yaml:7]: `pull-publishing-bot-validate` is now green :tada:

**Comment by wzshiming** [comment]
https://golang.org/dl/
go 1.6 has been released whether to use it

**Comment by justaugustus** [comment]
Rebased to remove the bazel-related commits.
Shoutout to @BenTheElder! 💕 
ref: https://github.com/kubernetes/kubernetes/pull/99561

**Comment by justaugustus** [comment]
The [dependencies test is failing](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-dependencies/1366533122371358720) on verify-vendor.sh, which I believe is a result of the go.mod/go.sum diffs between go1.15 and go1.16.

`gcr.io/k8s-testimages/kubekins-e2e:v20210226-c001921-master` is built on go1.15, while this PR should be tested w/ go1.16.

**Comment by justaugustus** [comment]
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

ref: https://github.com/kubernetes/test-infra/pull/21107

**Comment by justaugustus** [comment]
/test pull-kubernetes-dependencies-go-canary
/test pull-kubernetes-verify-go-canary

ref: https://github.com/kubernetes/test-infra/pull/21108

**Comment by liggitt** [comment]
all the bits in this PR lgtm

do we have a canary integration job that uses go1.16?

**Comment by justaugustus** [comment]
> do we have a canary integration job that uses go1.16?

@liggitt -- Opened a PR to add this: https://github.com/kubernetes/test-infra/pull/21110.

**Comment by justaugustus** [comment]
I needed to fix up some staticcheck warnings in https://github.com/kubernetes/kubernetes/pull/98572/commits/d9435800b65d6787beaf061546599d757b8e87c9, but the following go1.16 presubmits are now passing:
- `pull-kubernetes-dependencies-go-canary`: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-dependencies-go-canary/1366634510489227264
- `pull-kubernetes-verify-go-canary`: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-verify-go-canary/1366631602737647616

I have some confidence that the remaining tests currently running (pull-kubernetes-e2e-azure-disk-windows, pull-kubernetes-e2e-azure-file-windows-containerd, pull-kubernetes-e2e-gce-100-performance, pull-kubernetes-e2e-gce-ubuntu-containerd) will pass.

Once they do, https://github.com/kubernetes/test-infra/pull/21109 will need to be approved and merged, which should get the `pull-kubernetes-dependencies` and `pull-kubernetes-verify` jobs to go green.

**Comment by justaugustus** [comment]
Or not 🙃 
/test pull-kubernetes-e2e-azure-disk-windows

**Comment by justaugustus** [comment]
https://github.com/kubernetes/test-infra/pull/21109 has merged:
/test pull-kubernetes-dependencies
/test pull-kubernetes-verify

**Comment by justaugustus** [comment]
All green now! ✨ 🌈

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Review by cpanato — APPROVED** [review]
/lgtm

**Comment by justaugustus** [comment]
`pull-kubernetes-integration-go-canary` passed as well: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98572/pull-kubernetes-integration-go-canary/1366759123974098944

**Comment by justaugustus** [comment]
Screenshot from 2021-03-02 10-56-22

/label wake-up-bots
/honk

**Pull request merged**